Reduce unnecessary work in VMEntryScope to speed up first VM entry.

MenloDorian opened the pull request:
#### 3b78627af7fadf9e392d0eb8a7bcdfa96e01a20d
<pre>
Reduce unnecessary work in VMEntryScope to speed up first VM entry.
<a href="https://bugs.webkit.org/show_bug.cgi?id=253412">https://bugs.webkit.org/show_bug.cgi?id=253412</a>
rdar://106258354

Reviewed by Yusuke Suzuki.

1. VMEntryScope entry and exit current check for the need to do a lot of work that are usually not needed.
   Instead of doing all these checks, we introduce a new VM::EntryScopeService concept where if one of
   these checks are needed, their sources will request an EntryScopeService.  With this, the VMEntryScope
   will only check if any such services have been requested (with a single bitfield), and can will only
   incur the cost of the check if at least one such services have been requested.

   The only exception to using these service requests is the resetting of the VM DateCache.  That is
   dependent on a global lastTimeZoneID because there is one DateCache per VM, and there can be multiple
   VMs.  Hence, lastTimeZoneID needs to be updated in an atomic way.  We also don&apos;t want the EntryScope
   service request to require a lock of any sort, nor have to deal with iterating all VMs to request a
   DateCache reset.  So, we&apos;ll just do a hasTimeZoneChange() check instead in addition to the service
   request check on entry.

2. Moved the servicing of the &quot;EntryScopeService&quot;s out to VM::executeEntryScopeServicesOnEntry() and
   VM::executeEntryScopeServicesOnExit() instead of keeping them inlined in VMEntryScope.

   This appears to have improved the microbenchmark result by ~2-4%.  Might be due to improved cache
   locality from moving the rare case out.

3. Rename VM::m_terminationInProgress to VM::m_hasTerminationRequest to better describe what it represents.
   I wrote this code, and I found even myself confused by it.  So, I&apos;m renaming it for clarity, and
   updating the comment about its treatment on VM exit for added clarity as well.

4. Added cpp-to-js-cached-call.js and cpp-to-js-call-as-first-entry.js microbenchmarks which tests
   repeated calls using CachedCall, and repeated calls at the first entry boundary.

The microbenchmark results are all follows,  As expected, only first entry is improved by this change.

                                          base                      new

    cpp-to-js-call-as-first-entry       96.8583+-0.1895     ^     75.9927+-0.6285        ^ definitely 1.2746x faster
    cpp-to-js-call                      69.6459+-0.1565           69.5432+-0.0586
    cpp-to-js-cached-call               79.8228+-0.5376     ?     80.4139+-0.6115        ?

* JSTests/microbenchmarks/cpp-to-js-cached-call.js: Added.
(test):
* JSTests/microbenchmarks/cpp-to-js-call-as-first-entry.js: Added.
(test):
* Source/JavaScriptCore/jit/JITOperations.cpp:
(JSC::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/runtime/JSDateMath.h:
(JSC::DateCache::hasTimeZoneChange):
(JSC::DateCache::resetIfNecessary):
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::VM):
(JSC::VM::primitiveGigacageDisabled):
(JSC::VM::ensureWatchdog):
(JSC::VM::ensureSamplingProfiler):
(JSC::VM::whenIdle):
(JSC::VM::setException):
(JSC::VM::throwTerminationException):
(JSC::VM::assertScratchBuffersAreEmpty):
(JSC::VM::executeEntryScopeServicesOnEntry):
(JSC::VM::executeEntryScopeServicesOnExit):
(JSC::VM::clearScratchBuffers): Deleted.
* Source/JavaScriptCore/runtime/VM.h:
(JSC::VM::hasTerminationRequest const):
(JSC::VM::clearHasTerminationRequest):
(JSC::VM::setHasTerminationRequest):
(JSC::VM::hasAnyEntryScopeServiceRequest):
(JSC::VM::isValidEntryScopeService):
(JSC::VM::hasEntryScopeServiceRequest):
(JSC::VM::clearEntryScopeService):
(JSC::VM::requestEntryScopeService):
(JSC::VM::hasTimeZoneChange):
(JSC::VM::notifyNeedTermination):
(JSC::VM::terminationInProgress const): Deleted.
(JSC::VM::setTerminationInProgress): Deleted.
(JSC::VM::firePrimitiveGigacageEnabledIfNecessary): Deleted.
(JSC::VM::invokeDidPopListeners): Deleted.
(JSC::VM::resetDateCacheIfNecessary): Deleted.
* Source/JavaScriptCore/runtime/VMEntryScope.cpp:
(JSC::VMEntryScope::setUpSlow):
(JSC::VMEntryScope::tearDownSlow):
* Source/JavaScriptCore/runtime/VMTraps.cpp:
(JSC::VMTraps::handleTraps):
(JSC::VMTraps::deferTerminationSlow):
(JSC::VMTraps::undoDeferTerminationSlow):
* Source/JavaScriptCore/runtime/VMTrapsInlines.h:
(JSC::VMTraps::undoDeferTermination):
* Source/JavaScriptCore/tools/JSDollarVM.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSDollarVM::finishCreation):

Canonical link: <a href="https://commits.webkit.org/261260@main">https://commits.webkit.org/261260@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3cb041d26a62e6b9ddc1b443c59b04c4dbff539a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/111079 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/20219 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/43643 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/2506 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/119923 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/115036 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/21593 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/11321 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/2144 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/116834 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/16037 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/99221 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/103584 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/97994 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/30886 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/44505 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/99658 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/12736 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/32219 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/86407 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/10841 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/13278 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/9202 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/100802 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/18695 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/51817 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/31436 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7809 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/15229 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/108839 "Built successfully") | 
| | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/26838 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/VM.h:
```diff
@@ -393,6 +406,43 @@ class VM : public ThreadSafeRefCounted<VM>, public DoublyLinkedListNode<VM> {
+        ResetTerminationRequest,
+    };
+
+    uint8_t m_entryScopeServices { 0 };
```
Let's use `OptionSet`.

MenloDorian replied:
Thanks.  Fixed.

Constellation commented on Source/JavaScriptCore/runtime/VM.cpp:
```diff
@@ -1371,12 +1379,14 @@ ScratchBuffer* VM::scratchBufferForSize(size_t size)
     Locker locker { m_scratchBufferLock };
     for (auto* scratchBuffer : m_scratchBuffers)
-        scratchBuffer->setActiveLength(0);
+        ASSERT(!scratchBuffer->activeLength());
```
Actually, we have a problem here. See https://trac.webkit.org/changeset/232490/webkit change's comment

MenloDorian replied:
Thanks for catching this.  I restored it and trigger the clean up using an `EntryScopeService`.

Constellation reviewed (approved):
r=me

Commit:
Reduce unnecessary work in VMEntryScope to speed up first VM entry.
https://bugs.webkit.org/show_bug.cgi?id=253412
rdar://106258354

Reviewed by Yusuke Suzuki.

1. VMEntryScope entry and exit current check for the need to do a lot of work that are usually not needed.
   Instead of doing all these checks, we introduce a new VM::EntryScopeService concept where if one of
   these checks are needed, their sources will request an EntryScopeService.  With this, the VMEntryScope
   will only check if any such services have been requested (with a single bitfield), and can will only
   incur the cost of the check if at least one such services have been requested.

   The only exception to using these service requests is the resetting of the VM DateCache.  That is
   dependent on a global lastTimeZoneID because there is one DateCache per VM, and there can be multiple
   VMs.  Hence, lastTimeZoneID needs to be updated in an atomic way.  We also don't want the EntryScope
   service request to require a lock of any sort, nor have to deal with iterating all VMs to request a
   DateCache reset.  So, we'll just do a hasTimeZoneChange() check instead in addition to the service
   request check on entry.

2. Moved the servicing of the "EntryScopeService"s out to VM::executeEntryScopeServicesOnEntry() and
   VM::executeEntryScopeServicesOnExit() instead of keeping them inlined in VMEntryScope.

   This appears to have improved the microbenchmark result by ~2-4%.  Might be due to improved cache
   locality from moving the rare case out.

3. Rename VM::m_terminationInProgress to VM::m_hasTerminationRequest to better describe what it represents.
   I wrote this code, and I found even myself confused by it.  So, I'm renaming it for clarity, and
   updating the comment about its treatment on VM exit for added clarity as well.

4. Added cpp-to-js-cached-call.js and cpp-to-js-call-as-first-entry.js microbenchmarks which tests
   repeated calls using CachedCall, and repeated calls at the first entry boundary.

The microbenchmark results are all follows,  As expected, only first entry is improved by this change.

                                          base                      new

    cpp-to-js-call-as-first-entry       96.8583+-0.1895     ^     75.9927+-0.6285        ^ definitely 1.2746x faster
    cpp-to-js-call                      69.6459+-0.1565           69.5432+-0.0586
    cpp-to-js-cached-call               79.8228+-0.5376     ?     80.4139+-0.6115        ?

* JSTests/microbenchmarks/cpp-to-js-cached-call.js: Added.
(test):
* JSTests/microbenchmarks/cpp-to-js-call-as-first-entry.js: Added.
(test):
* Source/JavaScriptCore/jit/JITOperations.cpp:
(JSC::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/runtime/JSDateMath.h:
(JSC::DateCache::hasTimeZoneChange):
(JSC::DateCache::resetIfNecessary):
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::VM):
(JSC::VM::primitiveGigacageDisabled):
(JSC::VM::ensureWatchdog):
(JSC::VM::ensureSamplingProfiler):
(JSC::VM::whenIdle):
(JSC::VM::setException):
(JSC::VM::throwTerminationException):
(JSC::VM::assertScratchBuffersAreEmpty):
(JSC::VM::executeEntryScopeServicesOnEntry):
(JSC::VM::executeEntryScopeServicesOnExit):
(JSC::VM::clearScratchBuffers): Deleted.
* Source/JavaScriptCore/runtime/VM.h:
(JSC::VM::hasTerminationRequest const):
(JSC::VM::clearHasTerminationRequest):
(JSC::VM::setHasTerminationRequest):
(JSC::VM::hasAnyEntryScopeServiceRequest):
(JSC::VM::isValidEntryScopeService):
(JSC::VM::hasEntryScopeServiceRequest):
(JSC::VM::clearEntryScopeService):
(JSC::VM::requestEntryScopeService):
(JSC::VM::hasTimeZoneChange):
(JSC::VM::notifyNeedTermination):
(JSC::VM::terminationInProgress const): Deleted.
(JSC::VM::setTerminationInProgress): Deleted.
(JSC::VM::firePrimitiveGigacageEnabledIfNecessary): Deleted.
(JSC::VM::invokeDidPopListeners): Deleted.
(JSC::VM::resetDateCacheIfNecessary): Deleted.
* Source/JavaScriptCore/runtime/VMEntryScope.cpp:
(JSC::VMEntryScope::setUpSlow):
(JSC::VMEntryScope::tearDownSlow):
* Source/JavaScriptCore/runtime/VMTraps.cpp:
(JSC::VMTraps::handleTraps):
(JSC::VMTraps::deferTerminationSlow):
(JSC::VMTraps::undoDeferTerminationSlow):
* Source/JavaScriptCore/runtime/VMTrapsInlines.h:
(JSC::VMTraps::undoDeferTermination):
* Source/JavaScriptCore/tools/JSDollarVM.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSDollarVM::finishCreation):

Canonical link: https://commits.webkit.org/261260@main
